src/bom/bitcode.rs

```diff
@@ -637,59 +594,41 @@ fn build_bitcode_tar(bc_target : &OsString,
 /// We pass in the working directory in which the objects were constructed so
 /// that we can generate appropriate commands (in terms of absolute paths) so
 /// that we don't need to worry about where we are replaying the build from.
-fn attach_bitcode(chan : &mut mpsc::Sender<Option<Event>>,
-                  cwd : &Path,
-                  orig_target : &OsString,
+fn attach_bitcode(cwd : &Path,
+                  bc_args : &mut BitcodeArguments,
                   bc_target : &OsString) -> anyhow::Result<()> {
-    let object_path = to_absolute(cwd, orig_target);
-    let bc_path = to_absolute(cwd, bc_target);
-
-    let mut hasher = Sha256::new();
-    let mut bc_content = Vec::new();
-    let mut bc_file = std::fs::File::open(&bc_path)?;
-    bc_file.read_to_end(&mut bc_content)?;
-    hasher.update(bc_content);
-    let hash = hasher.finalize();
-
-    // This temporary file is filled in by `build_bitcode_tar`; however, we
-    // allocate it here so that it is still in scope (i.e., not deleted) when we
-    // invoke `inject_bitcode`
-    let mut tar_file = tempfile::NamedTempFile::new()?;
-    build_bitcode_tar(bc_target, bc_path.as_path(), hash.as_slice(), &mut tar_file)?;
-
-    let tar_name = OsString::from(tar_file.path());
-    let ok_tar_name = tar_name.into_string().ok().unwrap();
-    let opath = object_path.into_os_string();
-    inject_bitcode(chan, cwd, orig_target, bc_target, &ok_tar_name, &opath)
-}
-
-pub const ELF_SECTION_NAME : &str = ".llvm_bitcode";
+    let mktar = bc_args.ops.push_op(
+        SubProcOperation::calling(
+            |_cwd, args|
+            build_bitcode_tar(&args[0],
+                              &PathBuf::from(args[2].clone()).as_path(),
+                              &args[1],
+                              File::create(PathBuf::from(args[3].clone()).as_path())?)));
```

Comment:
I had a visceral reaction to seeing this API used in action. It seems like we absolutely must match the order of arguments (as they are pushed below) here. That feels distressingly error prone.

Is this `calling` primitive here so that we can re-arrange arguments and transform them after they are specified?

Reply:
I agree that there is a brittleness here, but I think it's intrinsic to the general functionality being exercised.  

Normally, a `SubProcOperation` identifies an executable to invoke and specifies some command-line parameters (input and output files) for that executable, as well as some arguments.  That executable, when invoked, will parse that command line to reconstruct its internal model of those arguments and parameters.

The `ChainedSubOps` takes a number of these `SubProcOperations` and chains them together in a sequence.  In particular, it uses the output file of one `SubProcOperation` in the chain as the input file of the subsequent `SubProcOperation` in the chain, handling this automatically for the client constructing the chain.  In many cases, a `TempFile` can indicate that a temporary file is to be used for those files and `ChainedSubOps` will automatically create and manage those temporary files.

Sometimes, instead of invoking an external executable, it is desireable to perform local functionality as a step in the chain. That is the case here where clang has generated the bitcode and internal code is being used to create a tarfile to contain that bitcode (the next step in the chain will insert that tarfile into the ELF file via `objcopy`.  So here, `calling` is saying "call this internal function" instead of "run this executable", so the "command line" parsing is happening directly through references to the args.

One alternative would be to split the chain at that point, execute the first chain, collect the results and run the local operation, then initiate the second chain, but that seemed to increase the overall complexity for the user.

That said, I'm realizing there is a phase issue here: the code generating the hash will be attempting to access the bitcode file before it has been generated because the chain hasn't been executed yet.  That will move some of the functionality into the `calling` target and reduce the number of overall arguments, but the actual files will still need to be arguments since they are temporaries.  I will need to make some additional changes (embarrassing)  and will update this PR soon from that.